include/storage/serialization.hpp

```diff
@@ -29,160 +33,190 @@ namespace serialization
         value = (value << 1) | data[index];
     return value;
 }
 
-template <typename T>
-inline void unpackBits(T &data, std::size_t index, std::size_t count, unsigned char value)
+template <typename T, typename BlockT = unsigned char>
+inline void unpackBits(T &data, std::size_t index, std::size_t count, BlockT value)
 {
     static_assert(std::is_same<typename T::value_type, bool>::value, "value_type is not bool");
-    const unsigned char mask = 1 << (count - 1);
+    const BlockT mask = 1 << (count - 1);
     for (std::size_t bit = 0; bit < count; value <<= 1, ++bit, ++index)
         data[index] = value & mask;
 }
 
-template <> inline void read<bool>(io::FileReader &reader, util::vector_view<bool> &data)
+template <typename VectorT>
+void readBoolVector(tar::FileReader &reader, const std::string &name, VectorT &data)
 {
-    const auto count = reader.ReadElementCount64();
-    BOOST_ASSERT(data.size() == count);
+    const auto count = reader.ReadElementCount64(name);
+    data.resize(count);
     std::uint64_t index = 0;
-    for (std::uint64_t next = CHAR_BIT; next < count; index = next, next += CHAR_BIT)
-    {
-        unpackBits(data, index, CHAR_BIT, reader.ReadOne<unsigned char>());
-    }
-    if (count > index)
-        unpackBits(data, index, count - index, reader.ReadOne<unsigned char>());
+
+    constexpr std::uint64_t WORD_BITS = CHAR_BIT * sizeof(std::uint64_t);
+
+    const auto decode = [&](const std::uint64_t block) {
+        auto read_size = std::min<std::size_t>(count - index, WORD_BITS);
+        unpackBits(data, index, read_size, block);
+        index += WORD_BITS;
+    };
+
+    reader.ReadStreaming<std::uint64_t>(name, boost::make_function_output_iterator(decode));
 }
 
-template <> inline void write<bool>(io::FileWriter &writer, const util::vector_view<bool> &data)
+template <typename VectorT>
+void writeBoolVector(tar::FileWriter &writer, const std::string &name, const VectorT &data)
 {
     const auto count = data.size();
-    writer.WriteElementCount64(count);
+    writer.WriteElementCount64(name, count);
     std::uint64_t index = 0;
-    for (std::uint64_t next = CHAR_BIT; next < count; index = next, next += CHAR_BIT)
-    {
-        writer.WriteOne<unsigned char>(packBits(data, CHAR_BIT * index, CHAR_BIT));
-    }
-    if (count > index)
-        writer.WriteOne<unsigned char>(packBits(data, index, count - index));
+
+    constexpr std::uint64_t WORD_BITS = CHAR_BIT * sizeof(std::uint64_t);
+
+    // FIXME on old boost version the function_input_iterator does not work with lambdas
+    // so we need to wrap it in a function here.
+    const std::function<std::uint64_t()> encode_function = [&]() -> char {
+        auto write_size = std::min<std::size_t>(count - index, WORD_BITS);
+        auto packed = packBits<VectorT, std::uint64_t>(data, index, write_size);
+        index += WORD_BITS;
+        return packed;
+    };
+
+    std::uint64_t number_of_blocks = std::ceil((double)count / WORD_BITS);
```

Comment:
`(count + WORD_BITS - 1) / WORD_BITS` would be better to avoid floating-point arithmetic

Reply:
Ah nice trick. I know a few places were this would come in useful.